compiler/optimizer/LoopVersioner.cpp

```diff
@@ -4542,56 +4406,190 @@ void TR_LoopVersioner::versionNaturalLoop(TR_RegionStructure *whileLoop, List<TR
       for (; search.hasTreeTop(); search.advance())
          {
          TR::TreeTop *tt = search.currentTreeTop();
-         if (removedNodes.contains(tt->getNode()))
+         TR::Node *ttNode = tt->getNode();
+         if (removedNodes.contains(ttNode))
             continue;
 
-         if (comp()->isPotentialOSRPoint(tt->getNode(), NULL, true))
+         if (ttNode->getOpCodeValue() == TR::treetop
+             || ttNode->getOpCode().isNullCheck()
+             || ttNode->getOpCode().isResolveCheck())
             {
-            if (trace())
-               traceMsg(comp(), "safeToRemoveOSRGuards false due to potential OSR point at n%dn\n", tt->getNode()->getGlobalIndex());
-            safeToRemoveOSRGuards = false;
-            break;
+            TR::Node *child = ttNode->getChild(0);
+            if (child->getOpCode().isFunctionCall())
+               {
+               newPrivatizationOK = false;
+               if (_curLoop->_privatizationOK)
+                  traceCannot("privatize", child, comp());
+               }
             }
-         else if (tt->getNode()->isOSRGuard())
+
+         // If the VM is configured to allow for OSR-HCR, then HCR invalidation
+         // can only happen at a subset of GC points even in compilations using
+         // traditional HCR. We could try to take advantage of that here, but
```

Comment:
Some clarification (perhaps via more comments in this part of the code) would help me on these points. I didn't completely understand the connection between OSR-HCR and traditional HCR (I thought traditional HCR meant no OSR usage for HCR). This might also help me understand why we only need to do `newHCRGuardVersioningOK = false;` if the HCR mode is traditional.

Reply:
I think if I were to add code comments, they'd end up something like "In OSR-HCR, HCR can occur at OSR yield points" and "In traditional HCR, HCR can occur at any GC point," which is basically just restating what the code says in this case.

Reply:
Okay I wasn't thinking of the case when traditional HCR can be used in certain compilations with OSR-HCR in effect in general but now that I'm thinking about that, the existing comment makes more sense to me and the following code is also just setting whether its okay to do certain versioning transformations, regardless of the HCR mode in use for the compilation.